src/kademlia/node.cpp

```diff
@@ -1167,12 +1185,38 @@ void node::incoming_request(msg const& m, entry& e)
 		}
 
 		sha1_hash target(target_ent.string_ptr());
-		nodes_t n;
 		// always return nodes as well as peers
-		m_table.find_node(target, n, 0);
-		write_nodes_entry(reply, n);
+		write_nodes_entries(target, arg_ent.dict_find_list("want"), reply);
 		return;
 	}
 }
 
+void node::write_nodes_entries(sha1_hash const& info_hash
+	, bdecode_node const& want, entry& r)
+{
+	// if no wants entry was specified, include a nodes
+	// entry based on the protocol the request came in with
+	if (want.type() != bdecode_node::list_t)
+	{
+		nodes_t n;
+		m_table.find_node(info_hash, n, 0);
+		write_nodes_entry(r[native_nodes_key()], n);
+		return;
+	}
+
+	for (int i = 0; i < want.list_size(); ++i)
```

Comment:
m_nodes is a reference to a std::map in dht_tracker which maps the "n4" and "n6" names to their respective dht::node objects.